File: Data Structures & Algorithms/trapping-rain-water/submission-3.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        if (height.size() < 3) return 0;
        
        int leftMax = height[0], rightMax = height[height.size()-1];
        int left = 1, right = height.size()-2;
        int res = 0;

        while (left <= right){
            if (rightMax <= leftMax){
                res += max(0, rightMax - height[right]);
                rightMax = max(rightMax, height[right]);
                right--;
            }
            else{
                res += max(0, leftMax - height[left]);
                leftMax = max(leftMax, height[left]);
                left++;
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.cs
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int res = 0, l = 0;
        var dict = new Dictionary<char,int>();

        for (int r = 0; r < s.Length; r++){
            if (dict.TryGetValue(s[r], out var val)){
                l = Math.Max(l, val + 1);
            }

            dict[s[r]] = r;
            res = Math.Max(res, r-l+1);
        }

        return res;
    }
}
